### PR TITLE
Tests for jp2image -> fix #739

### LINKED ISSUE #739: [bug report] User-controlled heap-based buffer overflow in exiv2 0.27.99.0, Exiv2::Jp2Image::encodeJp2Header(Exiv2::DataBuf const&, Exiv2::DataBuf&)  in jp2image.cpp:673
Hi,
I found there is a heap-based buffer overflow in exiv2 0.27.99.0, `Exiv2::Jp2Image::encodeJp2Header(Exiv2::DataBuf const&, Exiv2::DataBuf&)`  in `jp2image.cpp:673`, which will lead to DoS or code execution.

Test this [POC](https://drive.google.com/file/d/1EJJKmE8X5q783Fu4UQMwaC_BKsaAgLzf/view?usp=sharing) via 
> exiv2 rm /path/to/poc

The ASAN informations is as follows:
### ASAN
```
=================================================================
==28261==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60300000ec34 at pc 0x7febe838e935 bp 0x7ffd5b046cc0 sp 0x7ffd5b046468
READ of size 1785737760 at 0x60300000ec34 thread T0
    #0 0x7febe838e934 in __asan_memcpy (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x8c934)
    #1 0x7febe7eb21a5 in Exiv2::Jp2Image::encodeJp2Header(Exiv2::DataBuf const&, Exiv2::DataBuf&) /home/ubuntu/exiv2/src/jp2image.cpp:673
    #2 0x7febe7eb2815 in Exiv2::Jp2Image::doWriteMetadata(Exiv2::BasicIo&) /home/ubuntu/exiv2/src/jp2image.cpp:784
    #3 0x7febe7eb1d51 in Exiv2::Jp2Image::writeMetadata() /home/ubuntu/exiv2/src/jp2image.cpp:607
    #4 0x469df5 in Action::Erase::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/home/ubuntu/test/debug/exiv2-asan+0x469df5)
    #5 0x417c08 in main (/home/ubuntu/test/debug/exiv2-asan+0x417c08)
    #6 0x7febe670c82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #7 0x416578 in _start (/home/ubuntu/test/debug/exiv2-asan+0x416578)

0x60300000ec34 is located 0 bytes to the right of 20-byte region [0x60300000ec20,0x60300000ec34)
allocated by thread T0 here:
    #0 0x7febe839b6b2 in operator new[](unsigned long) (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x996b2)
    #1 0x7febe7edea0c in Exiv2::DataBuf::DataBuf(long) /home/ubuntu/exiv2/src/types.cpp:141
    #2 0x7febe7eb26d1 in Exiv2::Jp2Image::doWriteMetadata(Exiv2::BasicIo&) /home/ubuntu/exiv2/src/jp2image.cpp:759
    #3 0x7febe7eb1d51 in Exiv2::Jp2Image::writeMetadata() /home/ubuntu/exiv2/src/jp2image.cpp:607
    #4 0x469df5 in Action::Erase::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/home/ubuntu/test/debug/exiv2-asan+0x469df5)
    #5 0x417c08 in main (/home/ubuntu/test/debug/exiv2-asan+0x417c08)
    #6 0x7febe670c82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-buffer-overflow ??:0 __asan_memcpy
Shadow bytes around the buggy address:
  0x0c067fff9d30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9d40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9d50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9d60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9d70: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c067fff9d80: fa fa fa fa 00 00[04]fa fa fa 00 00 00 fa fa fa
  0x0c067fff9d90: fd fd fd fa fa fa fd fd fd fa fa fa fd fd fd fa
  0x0c067fff9da0: fa fa fd fd fd fa fa fa fd fd fd fa fa fa 00 00
  0x0c067fff9db0: 04 fa fa fa 00 00 00 fa fa fa 00 00 00 07 fa fa
  0x0c067fff9dc0: 00 00 04 fa fa fa 00 00 00 00 fa fa fd fd fd fa
  0x0c067fff9dd0: fa fa fd fd fd fa fa fa fd fd fd fa fa fa fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
==28261==ABORTING
```

And the reason seems to be the size of `output.pData_` and `subBox.length` depend on box header of the jp2 image, and exiv2 doesn't do a check before a memcpy.

### PR DESCRIPTION
**Pull request by D4N**
work in progress:
- [ ] add tests for jp2image, with good line & branch coverage
- [ ] fix the unaligned reads that were hidden because the warning was explicitly silenced
- [ ] fix #739

Note: the tests will probably fail badly, as UBSAN aborts the tests as soon as you touch `Jp2Image::writeMetadata()`.

### CODE REVIEW

File: unitTests/test_jp2image.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+#include "exiv2/jp2image.hpp"
+
+using namespace Exiv2;
+
+struct FakeJp2Image
+{
+    // static const byte signature_[];
+};
+
+/// copy pasta from jp2image.cpp
```
Comment (piponazo) [inline unitTests/test_jp2image.cpp]: pasta! 🍝

File: unitTests/test_jp2image.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x01, 0x07, 0x01, 0x01, 0xff, 0x64, 0x00, 0x23, 0x00, 0x01, 0x43, 0x72, 0x65, 0x61, 0x74, 0x6f, 0x72, 0x3a,
+    0x20, 0x4a, 0x61, 0x73, 0x50, 0x65, 0x72, 0x20, 0x56, 0x65, 0x72, 0x73, 0x69, 0x6f, 0x6e, 0x20, 0x31, 0x2e,
+    0x39, 0x30, 0x30, 0x2e, 0x31, 0xff, 0x52, 0x00, 0x0c, 0x00, 0x00, 0x00, 0x01, 0x00, 0x05, 0x04, 0x04, 0x00,
+    0x01, 0xff, 0x5c, 0x00, 0x13, 0x40, 0x40, 0x48, 0x48, 0x50, 0x48, 0x48, 0x50, 0x48, 0x48, 0x50, 0x48, 0x48,
+    0x50, 0x48, 0x48, 0x50, 0xff, 0x90, 0x00, 0x0a, 0x00, 0x00, 0x00, 0x00, 0x00, 0x2d, 0x00, 0x01, 0xff, 0x5d,
+    0x00, 0x14, 0x00, 0x40, 0x40, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0xff, 0x93, 0xcf, 0xb4, 0x04, 0x00, 0x80, 0x80, 0x80, 0x80, 0x80, 0xff, 0xd9};
+
+struct EncodeJp2Header : ::testing::Test
```
Comment (piponazo) [inline unitTests/test_jp2image.cpp]: This is not used, and few other things in this file. I guess you were experimenting and forgot to remove these leftovers, right?

File: unitTests/test_jp2image.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+{
+    Jp2Image img(this->mockIo_, true);
+    img.io().seek(0, BasicIo::beg);
+    for (size_t i = 0; i < EXV_COUNTOF(Jp2Blank); ++i) {
+        ASSERT_EQ(Jp2Blank[i], img.io().getb());
+    }
+    ASSERT_EQ(img.io().getb(), EOF);
+}
+
+TEST_F(aJp2Image, writeFailsOnClosedIo)
```
Comment (piponazo) [inline unitTests/test_jp2image.cpp]: The test name here should mention that `write` is throwing because the input file is not really a Jp2 image.

**Comment by clanmills** [comment]
There's an issue in `Jp2Image::printStructure()` causing an exception when we dump the header structure.  
```
$ for i in test/data/*.jp2 ; do exiv2 -pS $i; done
STRUCTURE OF JPEG2000 FILE: test/data/Reagan.jp2
 address |   length | box       | data
       0 |       12 | jP        | 
      12 |       20 | ftyp      | 
      32 |       45 | jp2h      | 
      40 |       22 |  sub:ihdr | Uncaught exception: Invalid slice bounds specified
STRUCTURE OF JPEG2000 FILE: test/data/Reagan2.jp2
 address |   length | box       | data
       0 |       12 | jP        | 
      12 |       20 | ftyp      | 
      32 |     3185 | jp2h      | 
      40 |       22 |  sub:ihdr | Uncaught exception: Invalid slice bounds specified
STRUCTURE OF JPEG2000 FILE: test/data/relax.jp2
 address |   length | box       | data
       0 |       12 | jP        | 
      12 |       20 | ftyp      | 
      32 |      345 | jp2h      | 
      40 |       22 |  sub:ihdr | Uncaught exception: Invalid slice bounds specified
706 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ 
```

This appears to be caused by the '30' passed to makeSlice() when there are less than 30 bytes in data.buf_

My fix is:
```
706 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ git diff
diff --git a/src/jp2image.cpp b/src/jp2image.cpp
index 39a6f58..80d1ec2 100644
--- a/src/jp2image.cpp
+++ b/src/jp2image.cpp
@@ -508,9 +508,11 @@ namespace Exiv2
                             DataBuf data(subBox.length-sizeof(box));
                             io_->read(data.pData_,data.size_);
                             if ( bPrint ) {
+                                size_t ssize = data.size_;
+                                if   ( ssize > 30 ) ssize = 30; // limit the binary data presentation to 30 bytes
                                 out << Internal::stringFormat("%8ld | %8ld |  sub:",(size_t)address,(size_t)subBox.length)
                                     << toAscii(subBox.type)
-                                    <<" | " << Internal::binaryToString(makeSlice(data, 0, 30));
+                                    <<" | " << Internal::binaryToString(makeSlice(data, 0, ssize));
                                 bLF = true;
                             }
 
707 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ 
```

With this fix, the output is:
```
.../exiv2/0.27-maintenance $ for i in test/data/*.jp2 ; do  cmake-build-debug/bin/exiv2 -pS $i; done
STRUCTURE OF JPEG2000 FILE: test/data/Reagan.jp2
 address |   length | box       | data
       0 |       12 | jP        | 
      12 |       20 | ftyp      | 
      32 |       45 | jp2h      | 
      40 |       22 |  sub:ihdr | .............
      62 |       15 |  sub:colr | ....... | pad: 1 0 0 | iccLength:16
      77 |     1342 | uuid      | Exif: II*...........................>........
    1419 |      942 | uuid      | IPTC: ............040621-N-6536T-062.....5...
    2361 |     5590 | uuid      | XMP : <?xpacket begin="..." id="W5M0MpCehiHzre
    7951 |    32658 | jp2c      | 
STRUCTURE OF JPEG2000 FILE: test/data/Reagan2.jp2
 address |   length | box       | data
       0 |       12 | jP        | 
      12 |       20 | ftyp      | 
      32 |     3185 | jp2h      | 
      40 |       22 |  sub:ihdr | .............
      62 |     3155 |  sub:colr | ......HLino....mntrRGB XYZ .. | pad: 2 0 0 | iccLength:3144
    3217 |        0 | jp2c      | 
STRUCTURE OF JPEG2000 FILE: test/data/relax.jp2
 address |   length | box       | data
       0 |       12 | jP        | 
      12 |       20 | ftyp      | 
      32 |      345 | jp2h      | 
      40 |       22 |  sub:ihdr | ...,.........
      62 |      289 |  sub:colr | ............ ..scnrRGB XYZ .. | pad: 2 0 0 | iccLength:278
     351 |       26 |  sub:res  | ....resc$I..$I....
     377 |        0 | jp2c      | 
708 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ 
```
This fault applies to both 'master' and '0.27-maintenance'.

File: unitTests/test_jp2image.cpp (hunk 2 above)
Comment (D4N) [reply unitTests/test_jp2image.cpp]: This is WIP for a reason ;-)

**Comment by D4N** [comment]
@clanmills Can you put that in a separate issue? I'd like to first focus on adding tests to the current behavior and not modify it beforehand.

**Comment by clanmills** [comment]
Sure, Dan.  I'll open a new issue.  I'll _somehow_ find the courage/strength to attempt a PR with my fix.

**Comment by clanmills** [comment]
I'm going to close this.  src/jpimage.cpp will be thoroughly reviewed in v1.00 in #1525

**Pull request closed without merging**